File: packages/app/src/sim/city.ts
// Loads the committed city manifest + demand grid and builds the core city JSON that
// Sim::new consumes. The demand grid is stored in lng/lat; we convert each cell to local
// mm HERE (coords/geo.ts boundary) so the sim only ever sees mm (snake_case keys match the
// Rust serde field names of DemandGrid/DemandCell — NOT camelCase).
import { lngLatToMm, setOrigin } from "../coords/geo";
import { withBase } from "../config";

export interface RawCity {
  id: string;
  name: string;
  originLngLat: [number, number];
  bbox: [number, number, number, number];
  center: [number, number];
  zoom: number;
  seed: number;
  demandGridPath: string;
  /** Which engine ruleset to construct (fantasy-fork.md): absent/"transit" = the classic game;
   *  "arcadia" = the hex 4X-logistics fantasy campaign. Mirrors `CityData.ruleset`. */
  ruleset?: string;
  /** Hex-lattice cell size (mm) for the fantasy map — mirrors `CityData.grid_cell_mm`. Absent/0 =
   *  the continuous Catmull-Rom geometry (every transit city). >0 builds track on the hex lattice. */
  gridCellMm?: number;
  /** Optional committed real-world starting network (e.g. the MRT). */
  networkPath?: string;
  /** Optional committed buildability grid (surface-rail cost signal). */
  buildabilityPath?: string;
  /** Fantasy (arcadia) supply graph baked by scripts/build_world.py — an ADDITIVE manifest field
   *  (serde-ignored elsewhere; never copied into the core city JSON). S2: resource nodes that fork the
   *  two supply chains. Positions carried as axial (q,r) + i64 mm (= hexgrid::center_of); yields i64. */
  supplyGraph?: {
    resources: { kind: string; q: number; r: number; xMm: number; yMm: number; yield: number }[];
    /** S3 towns: supply sinks + conquest targets. kind = "capital"|"starter"|"neutral"; value = i64
     *  conquest reward; demands = nearest resource kinds; decadence = S4 per-town corruption floor. */
    towns?: { kind: string; q: number; r: number; xMm: number; yMm: number; value: number; demands: string[]; decadence: number; recipe?: number[] }[];
    /** S4 decadence seed: the far-edge reservoir (tide origin + raider anchors), the clean grace radius,
     *  and the realm's baked STARTING decadence (seeded into world.decadence). */
    decadenceSeed?: { capitalGraceHexes: number; reservoir: { q: number; r: number; xMm: number; yMm: number }[]; initialDecadence?: number; growthPerS?: number; armySpeedMmS?: number; creepPerS?: number; productionMicro?: number; capitalXMm?: number; capitalYMm?: number; influenceHops?: number; initialGold?: number; buildGoldDivisor?: number; goldUpkeepPerDay?: number; manpowerUpkeepPerLegionDay?: number; rivalEnabled?: boolean; walkBackstopMicro?: number };
  };
  /** Additive baked DRAINAGE topology (build_world.py flow-accumulation rivers) — render-only; never copied
   *  into the core city JSON (the sim never sees it). Each edge is a cell-centre→cell-centre segment (i64 mm
   *  endpoints) with a width class (1..4) + a ford flag (a cheap headwater crossing). */
  rivers?: { q: number; r: number; toQ: number; toR: number; x0Mm: number; y0Mm: number; x1Mm: number; y1Mm: number; wclass: number; ford: boolean }[];
  /** Optional per-city rider patience (sim-ms) — overrides the core's default. The globe sets an
   *  air-scale value (90_000 = 45 clock-min): air travellers arrive for a departure rather than
   *  drifting off after one missed metro interval, so its pressure is CAPACITY (denied boardings,
   *  aircraft choice), not schedule impatience. */
  patienceMs?: number;
}

export interface RawDemand {
  cellM: number;
  bbox: [number, number, number, number];
  /** `commodity` (fantasy S7e): the Forge-Line commodity a source cell produces (ORE=0 default). */
  cells: { lon: number; lat: number; originWeight: number; destWeight: number; commodity?: number }[];
}

export interface RawBuildability {
  cellM: number;
  bbox: [number, number, number, number];
  cells: { lon: number; lat: number; c: number }[];
}

export interface LoadedCity {
  raw: RawCity;
  seed: number;
  coreCityJson: string;
  demandCellCount: number;
  buildability?: RawBuildability;
  /** Demand grid as lng/lat heat points (origin+dest weight) for the demand map layer. */
  demandHeat: { lng: number; lat: number; weight: number }[];
  /** Demand grid cell pitch (m) — sizes the demand-heat hexagons so they tile the grid. */
  demandCellM: number;
}

/** Pure: build the core (mm) city JSON from the raw manifest + lng/lat grids. */
export function buildCoreCity(
  raw: RawCity,
  demand: RawDemand,
  buildability?: RawBuildability,
): { json: string; cellCount: number } {
  const cells = demand.cells.map((c) => {
    const [x_mm, y_mm] = lngLatToMm([c.lon, c.lat]);
    // commodity (fantasy S7e): which Forge-Line good a source cell produces (omitted ⇒ 0=ORE, serde default).
    return { x_mm, y_mm, origin_w: c.originWeight, dest_w: c.destWeight, commodity: c.commodity ?? 0 };
  });
  const core: Record<string, unknown> = { seed: raw.seed, demand: { cell_m: demand.cellM, cells } };
  if (raw.patienceMs !== undefined) core.patience_ms = raw.patienceMs; // per-city demand knob (city.rs)
  if (raw.ruleset) core.ruleset = raw.ruleset; // the fantasy-fork seam (World::new selects the mode)
  if (raw.ruleset === "arcadia") core.force_single_track = true; // fantasy: all track SINGLE (readability + forces meets/signals)
  if (raw.gridCellMm) core.grid_cell_mm = raw.gridCellMm; // hex lattice for the fantasy map
  const dec = raw.supplyGraph?.decadenceSeed;
  if (dec?.initialDecadence) core.initial_decadence = dec.initialDecadence; // baked starting corruption (S4)
  if (dec?.growthPerS) core.decadence_growth_per_s = dec.growthPerS; // baked lose-meter fill rate (balance)
  if (dec?.armySpeedMmS) core.army_speed_mm_s = dec.armySpeedMmS; // baked legion march speed (continent scale)
  if (dec?.capitalXMm) core.capital_x_mm = dec.capitalXMm; // baked capital cell (S10 decadence-tide target)
  if (dec?.capitalYMm) core.capital_y_mm = dec.capitalYMm;
  if (dec?.creepPerS) core.decadence_creep_per_s = dec.creepPerS; // baked tide creep rate (S10b-2)
  if (dec?.productionMicro) core.production_micro = dec.productionMicro; // baked economy pace (S11 — snappy 3-channel flow)
  if (dec?.influenceHops) core.influence_hops = dec.influenceHops; // baked area-of-influence radius (#9 — build gate)
  if (dec?.initialGold) core.initial_gold = dec.initialGold; // baked starting gold treasury (#economy)
  if (dec?.buildGoldDivisor) core.build_gold_divisor = dec.buildGoldDivisor; // baked gold build-cost scale (#economy)
  if (dec?.goldUpkeepPerDay) core.gold_upkeep_per_day = dec.goldUpkeepPerDay; // baked per-day gold upkeep (#economy opex)
  if (dec?.manpowerUpkeepPerLegionDay) core.manpower_upkeep_per_legion_day = dec.manpowerUpkeepPerLegionDay; // baked per-legion-day manpower upkeep (#daynight)
  if (dec?.rivalEnabled) core.rival_enabled = dec.rivalEnabled; // #13: seed a baked rival realm at construction
  if (dec?.walkBackstopMicro) core.walk_backstop_micro = dec.walkBackstopMicro; // baked off-rail goods backstop (#11)
  if (buildability) {
    core.buildability = {
      cell_m: buildability.cellM,
      cells: buildability.cells.map((c) => {
        const [x_mm, y_mm] = lngLatToMm([c.lon, c.lat]);
        return { x_mm, y_mm, c: c.c };
      }),
    };
  }
  return { json: JSON.stringify(core), cellCount: cells.length };
}

/** Fetch the manifest + grids, set the session's coordinate origin, and assemble the core
 *  city JSON. setOrigin MUST happen before buildCoreCity so lng/lat -> mm is correct. */
export async function loadCity(manifestPath: string): Promise<LoadedCity> {
  const raw: RawCity = await (await fetch(withBase(manifestPath))).json();
  setOrigin(raw.originLngLat[0], raw.originLngLat[1]);
  const demand: RawDemand = await (await fetch(withBase(raw.demandGridPath))).json();
  let buildability: RawBuildability | undefined;
  if (raw.buildabilityPath) {
    try {
      buildability = await (await fetch(withBase(raw.buildabilityPath))).json();
    } catch {
      buildability = undefined; // graceful: no penalties without the grid
    }
  }
  const { json, cellCount } = buildCoreCity(raw, demand, buildability);
  const demandHeat = demand.cells.map((c) => ({ lng: c.lon, lat: c.lat, weight: c.originWeight + c.destWeight }));
  return { raw, seed: raw.seed, coreCityJson: json, demandCellCount: cellCount, buildability, demandHeat, demandCellM: demand.cellM };
}

File: crates/sim/src/city.rs
//! The city manifest the core deserializes from `Sim::new(seed, city_json)`. The core
//! reads ONLY what it needs in sim-space (the demand grid in mm + seed); the frontend's
//! copy of CityData additionally carries lng/lat origin/bbox/center for the map. Unknown
//! JSON fields (originLngLat, bbox, ...) are ignored by serde, keeping lng/lat out of the
//! core. A future GTFS importer maps into the same types — additive, not a refactor.
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
pub struct CityData {
    #[serde(default)]
    pub id: String,
    /// Which ruleset the engine constructs for this city — the fantasy-fork seam
    /// (fantasy-fork.md). Frozen at construction, NEVER a Command. `"transit"` (the serde
    /// default for any city JSON omitting it) selects the classic gravity/agent transit game;
    /// future `"arcadia"` selects the hex 4X-logistics ruleset. `CityData::default()` leaves it
    /// empty (`""`), which `World::new` treats identically to `"transit"`, so every existing
    /// city and native test is byte-identical (zero re-pins). NOT part of `state_hash` — it is a
    /// construction-time selector, not evolving state.
    #[serde(default = "default_ruleset")]
    pub ruleset: String,
    #[serde(default)]
    pub seed: u64,
    /// Demand grid in sim millimetres (the frontend converts lon/lat -> mm before embedding).
    #[serde(default)]
    pub demand: DemandGrid,
    /// Coarse buildability grid (mm) for the surface-rail cost signal. Additive/optional.
    #[serde(default)]
    pub buildability: BuildabilityGrid,
    /// GRID geometry mode (fantasy-fork.md §10 / shared-rail.md). When > 0, track is built on a
    /// `grid_cell_mm` lattice (crisp octilinear, integer-exact) instead of the continuous Catmull-Rom
    /// curve, so two lines over the same physical corridor produce byte-identical edges (the
    /// foundation for cross-line shared track). A bake property frozen at construction, NOT a Command.
    /// **0 = off (continuous)** ⇒ every existing city builds the exact same geometry (zero re-pins).
    #[serde(default)]
    pub grid_cell_mm: i64,
    /// Fantasy (arcadia): the baked STARTING decadence — the ambient corruption the realm begins with,
    /// seeded by `scripts/build_world.py` S4 from the per-town floors (a more-corrupt continent starts
    /// further up the lose meter ⇒ more urgency). A bake property frozen at construction, NOT a Command;
    /// `World::new` seeds `world.decadence` from it. **0 (the serde + `Default` value) ⇒ the realm starts
    /// clean** — every existing city, the arcadia golden fixture, and every native test stay byte-identical
    /// (zero re-pins); only a baked world that sets it starts corrupt.
    #[serde(default)]
    pub initial_decadence: i64,
    /// Fantasy (arcadia) S7e/balance: decadence GROWTH per sim-second (the lose-meter fill rate). A
    /// per-city knob (the large baked continent presses far gentler than the small demo). **0 (serde +
    /// `Default`) ⇒ the `decadence::BASE_GROWTH_PER_S` default**, so every existing city + the golden
    /// fixtures + native tests keep the shipped (demo) balance — byte-identical.
    #[serde(default)]
    pub decadence_growth_per_s: i64,
    /// Fantasy (arcadia) balance: legion MARCH speed in mm per sim-second (a legion's pace riding the
    /// rails). A per-city knob — the large baked continent (towns 60+ km from the capital) needs a far
    /// faster legion than the 1.5 km demo, or conquest can't reach a town before the rot overruns the
    /// realm. **0 (serde + `Default`) ⇒ the `army::ARMY_SPEED_MM_S` default**, so the demo, the arcadia
    /// golden fixture, and every native test keep the shipped (demo) balance — byte-identical (with the
    /// default the marching `s_mm` trajectory is unchanged; only a baked world that sets it diverges).
    #[serde(default)]
    pub army_speed_mm_s: i64,
    /// Fantasy (arcadia) S10: the baked CAPITAL cell in sim mm (the seat the decadence tide races toward
    /// — the lose target + the creep-gradient origin). Seeded by `scripts/build_world.py` from the
    /// capital town. **(0, 0) (the serde + `Default` value) ⇒ no capital ⇒ no decadence CA** — every
    /// transit city, the arcadia golden fixture, and native tests build an empty `DecadenceField`
    /// (golden-neutral; the field is static/un-hashed anyway). A pure construction-time map property,
    /// NOT a Command.
    #[serde(default)]
    pub capital_x_mm: i64,
    #[serde(default)]
    pub capital_y_mm: i64,
    /// How long (sim ms) a waiting rider tolerates before giving up (renege). A per-city
    /// demand knob, NOT a hardcoded constant. Cities loaded from JSON without the field get
    /// `default_patience_ms`; `CityData::default()` leaves it 0, which DISABLES renege (handy
    /// for native tests that don't want abandonment).
    #[serde(default = "default_patience_ms")]
    pub patience_ms: i64,
    /// Maximum legs (transfers + 1) a routed trip may use. 0 (the `Default` value) means "use
    /// the routing default" — so it's a per-city knob a future RAPTOR can raise without a core
    /// change, while `CityData::default()` keeps the shipped behaviour.
    #[serde(default)]
    pub max_legs: usize,
    /// Demand growth per in-game day, in basis points (250 = +2.5%), applied to cells within a
    /// catchment of a SERVED station — transit-oriented development: the city grows where (and
    /// because) you serve it. Cells outside any catchment grow at a third of this rate (ambient
    /// sprawl), so a network you stop extending slowly falls behind the city. The pressure this
    /// creates is the one-more-day engine: growth is good news (riders, coverage) that creates
    /// problems (queues, crush load). Cities loaded from JSON without the field get
    /// `default_growth_bp`; `CityData::default()` leaves it 0, which DISABLES growth (native
    /// tests opt in explicitly).
    #[serde(default = "default_growth_bp")]
    pub growth_bp_per_day: i64,
}

/// Default transit-adjacent demand growth: +2.5% per in-game day (ambient = a third of this).
fn default_growth_bp() -> i64 {
    250
}

/// The ruleset a city JSON selects when it omits `ruleset` — the classic transit game. Shared by
/// both `CityData` and `SaveGame` so a save and the city it was played on agree on the seam. Note
/// `CityData::default()` (the native-test constructor) still yields `""` via `#[derive(Default)]`;
/// `World::new` canonicalises `"" | "transit"` to the transit ruleset, so both spell the same game.
pub fn default_ruleset() -> String {
    "transit".to_string()
}

/// Default rider patience for cities that don't specify one: 10 CLOCK-minutes (20_000 sim-ms in
/// the unified frame) — about two missed trains at a typical 3–6 clock-minute headway. Patience
/// arms the renege ("gave up waiting") pressure signal, the game's primary difficulty source; a
/// patience several times the longest sane headway would silently disable it.
fn default_patience_ms() -> i64 {
    20_000
}

/// Coarse classified grid: each cell carries a class `c` (1=RoadROW 2=RailROW 3=Built
/// 4=Water 5=Park; absent cells are Open=0). Built offline from OSM (scripts/build_buildability.py).
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
pub struct BuildabilityGrid {
    #[serde(default)]
    pub cell_m: f64,
    #[serde(default)]
    pub cells: Vec<BuildCell>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct BuildCell {
    pub x_mm: i64,
    pub y_mm: i64,
    pub c: u8,
}

/// Surface-rail cost class codes (shared by the grid + the sim).
pub mod class {
    pub const OPEN: u8 = 0;
    pub const ROAD: u8 = 1;
    pub const RAIL: u8 = 2;
    pub const BUILT: u8 = 3;
    pub const WATER: u8 = 4;
    pub const PARK: u8 = 5;
}

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
pub struct DemandGrid {
    /// Grid cell size in metres (metadata; cells already carry mm positions).
    #[serde(default)]
    pub cell_m: f64,
    #[serde(default)]
    pub cells: Vec<DemandCell>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DemandCell {
    pub x_mm: i64,
    pub y_mm: i64,
    /// Trip-origin weight (residents) and trip-destination weight (jobs/POIs).
    pub origin_w: f32,
    pub dest_w: f32,
    /// Fantasy (arcadia) S7e: which Forge-Line commodity this cell produces/consumes (ORE=0, GRAIN=1,
    /// AETHER=2, FUEL=3 …). A source station's output commodity is the dominant origin-commodity of its
    /// captured cells. **0 (the serde + `Default` value) ⇒ ORE** — every transit city + the golden
    /// fixtures stay byte-identical (transit has no forge), so this is a transit-neutral addition.
    #[serde(default)]
    pub commodity: u8,
}

impl CityData {
    /// Parse from the JSON string passed across the wasm boundary.
    pub fn from_json(json: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(json)
    }
}
